api/src/main/java/org/pmiops/workbench/initialcredits/InitialCreditsExpirationServiceImpl.java

```diff
@@ -129,51 +142,58 @@ public void extendInitialCreditsExpiration(DbUser user) {
   private void checkExpiration(DbUser user) {
     DbUserInitialCreditsExpiration userInitialCreditsExpiration =
         user.getUserInitialCreditsExpiration();
-    if (null != userInitialCreditsExpiration
-        && haveCreditsExpired(user)
-        && userInitialCreditsExpiration
-            .getNotificationStatus()
-            .equals(NotificationStatus.NO_NOTIFICATION_SENT)) {
-      logger.info(
-          "Initial credits expired for user {}. Expiration time: {}",
-          user.getUsername(),
-          userInitialCreditsExpiration.getExpirationTime());
-      boolean hasExhaustedWorkspace =
-          workspaceDao.findAllByCreator(user).parallelStream()
-              .anyMatch(DbWorkspace::isInitialCreditsExhausted);
-
-      // Set initial credits expired for all workspaces regardless of whether
-      // they have exhausted their credits or not.
-      workspaceDao.findAllByCreator(user).stream()
-          .filter(
-              ws ->
-                  BillingUtils.isInitialCredits(
-                      ws.getBillingAccountName(), workbenchConfigProvider.get()))
-          .filter(DbWorkspace::isActive)
-          .filter(ws -> !ws.isInitialCreditsExpired())
-          .forEach(
-              ws -> {
-                ws.setInitialCreditsExpired(true);
-                ws.setBillingStatus(BillingStatus.INACTIVE);
-                workspaceDao.save(ws);
-                deleteAppsAndRuntimesInWorkspace(ws);
-              });
-      try {
-        // If the user has already been notified about exhausting their initial credits,
-        // we do not need to notify them about expiration as well.
-        if (!hasExhaustedWorkspace) {
-          mailService.alertUserInitialCreditsExpired(user);
-          userInitialCreditsExpiration.setNotificationStatus(
-              NotificationStatus.EXPIRATION_NOTIFICATION_SENT);
-          userDao.save(user);
-        }
-
-      } catch (MessagingException e) {
-        logger.error(
-            String.format(
-                "Failed to send initial credits expiration notification for user %s",
-                user.getUserId()));
-      }
+    if (null == userInitialCreditsExpiration
+        || userInitialCreditsExpiration.getExpirationCleanupTime() != null) {
```

Comment:
I understand we have a cloud task that processes all users and we check for expirationCleanupTime here. Does it make sense to exclude these users earlier, I mean in the db query that gets the users instead of passing in users to the cloud task that we know we won't process?

Reply:
I like that idea. I will work on that and get back to you.

Reply:
I updated the controller to handle this. Please let me know if that seems like a good approach to you.
https://github.com/all-of-us/workbench/pull/8900/files#r1831277139